### ISSUE TITLE
wasm-merge fills in null data based on dylink section

### ISSUE THREAD

**Issue by AmaanC** [WebAssembly/binaryen#1271]
When trying to use `wasm-merge` to merge 2 WASM files:

1. One with a custom `dylink` section (such as ones generated by emscripten with `-s WASM=1 -s SIDE_MODULE=1`)
2. Second with no data section

The result of `wasm-merge one.wasm two.wasm -o merge.wast -S` is a WAST such as:

```
(module
 (type $0 (func))
 ...
 (type $3 (func))
 (import "env" "memoryBase" (global $import$0 i32))
 (import "env" "tableBase" (global $import$1 i32))
 (import "env" "table" (table 10 anyfunc))
 (import "env" "memoryBase" (global $import$0$0 i32))
 (import "env" "tableBase" (global $import$3 i32))
 (global $global$0 (mut i32) (i32.const 0))
 (global $global$1 (mut i32) (i32.const 0))
 
 (memory $0 80)
 ;; This is the issue
 (data (get_global $import$0) "\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00...
```

The number of `\00` bytes corresponds with the `memorysize` value in the `dylink` section (as [described here](https://github.com/WebAssembly/tool-conventions/blob/master/DynamicLinking.md#the-dylink-section)).

Regarding the `dylink` section:

> If the dynamic library has memorysize > 0 then the loader will reserve room in memory of that size and initialize it to zero (note: can be larger than the memory segments in the module, if the dynamic library wants additional space) at offset env.memoryBase, and similarly for the table (where initialization is to null, i.e., a trap will occur if it is called). The library can then place memory and table segments at the proper locations using those imports.

Should `wasm-merge` have an option to skip this behavior and leave it up to the actual loader (such as the browser) to initialize that memory to 0, given that `memoryBase` may also be an import and undetermined at the time of merge?

----

This behavior does seem intentional from the source, to account for the case when a `dylink` section bumps up the memory requirement. Does it make sense to allow disabling the behavior, though? (It causes the WASM files to seem quite bloated with several MiB of null bytes.)

**Edit:** I see that `wasm-merge` doesn't copy over the `dylink` section, which explains why it also pads the memory section based on `dylink`. Should we consider merging the `dylink` sections? Would it make sense to have the `--no-dylink-padding` option I added in my PR still?

**Comment by kripken** [comment]
I think the underlying issue here is that wasm is limited in what can be used to decide where to place a segment - that offset can only be a constant, or a get_global. If it could receive an add of a constant and a global, then we could optimize out those zeros.

Until wasm adds that, though, the zeros are annoying but they do almost vanish when gzipping, so it's not too bad, I think?

**Comment by AmaanC** [comment]
> Until wasm adds that, though, the zeros are annoying but they do almost vanish when gzipping, so it's not too bad, I think?

I haven't played with many settings but it's not as good as you might hope. (156 byte wasm + 309 byte wasm = 5.3kb gzip.)

> If it could receive an add of a constant and a global, then we could optimize out those zeros.

I don't quite understand what you mean; not quite sure how `dylink` and `memoryBase` interact, sorry! Is there something more comprehensive than the docs on WebAssembly.org and the tool-conventions one I linked to?

Would copying the `dylink` section (when only one of the files has the section) into the output file make sense?

For context, I'm trying to add the `call_indirect` operator into a wasm file generated by `emscripten`.

**Comment by AmaanC** [comment]
@kripken Thought I'd ask again in case you missed it:

> Would copying the dylink section (when only one of the files has the section) into the output file make sense?

I can start work on a PR for that if you think we could get that merged upstream!

**Comment by kripken** [comment]
Yes, I think you're right - what we should do is merge the dylink sections of the merged modules. We'd need to handle the case of just one having it, or both. And then we can avoid the zeros.

**Comment by tlively** [comment]
The new wasm-merge does not do anything with the dylink section, so this is no longer relevant.
